### ISSUE TITLE
[Clang 19.1.4] [C23] VLA warning / usage when size is specified as member of static constexpr struct

### ISSUE THREAD

**Issue by florianmarkusse** [llvm/llvm-project#118331]
Hi there,

I recently upgraded my personal project to the new C23 standard when I saw that Clang 19.1.4 supports both `constexpr` and `#embed`. Both features have helped my project get rid of some manual things, thanks for that.

However, when I create a `static constexpr struct` and use its member to set the size of an array, Clang warns me about VLA usage. These come in 2 flavors. See this [godbolt](https://godbolt.org/z/jE3ecr5en). 
- Warning about `VLA` usage which is then folded into a constant array as an extension. 
- Waqning about `VLA` usage 

I will post the code here again:
```C
static constexpr struct
{   int SHIFT;
    int ENTRIES;
}
PageTableFormat = {.SHIFT = 9, .ENTRIES = (1ULL << PageTableFormat.SHIFT)}; // Emits '-Wgnu-folding-constant'

static int arraysOfThings[(enum {SIZE = PageTableFormat.ENTRIES})SIZE];

int testFunction() {
    int canIBeConstant[SIZE];

    return 3;
}

static constexpr int size = PageTableFormat.ENTRIES;
int testFunction2() {
    int canIBeConstant[size];

    return 3;
}

int testFunction3() {
    int canIBeConstant[PageTableFormat.ENTRIES]; // Emits '-Wvla'

    return 3;
}

static int new_arraysOfThings[PageTableFormat.ENTRIES]; // Emits ' -Wvla' and '-Wgnu-folding-constant'

// Using a constexpr like this does work as expected
static constexpr int THIS_DOES_WORK_SIZE = 10;

int testFunction4() {
    int iAmConstant[THIS_DOES_WORK_SIZE];

    return 3;
}

int constantArray[THIS_DOES_WORK_SIZE];
```

**Expected behavior**
- Clang compiles by turning the arrays into constant arrays and not VLAs. 
- Clang does not emit a warning about using a VLA `[-Wvla]` if the array is folded into a constant array [`-Wgnu-folding-constant`] 
- NOTE: Should Clang even emit the [`-Wgnu-folding-constant`] warning? According to [N3220](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3220.pdf) paragraph 5.6:

> An identifier that is ... declared with storage-class specifier constexpr and has an object type, is a named constant, as is a postfix expression that applies the . member access operator to a named constant of structure or union type, even recursively.

So, i think that even [`-Wgnu-folding-constant`] shouldn't be emitted since it is part of the standard, right?

**Affected version**
`Clang 19.1.0` and up

**How to Reproduce**
- Use `Clang 19.1.0` or newer
- Create a `static constexpr struct` and use the member variable as the size of an array in a function or as an array  inside a translation unit. and compile

When creating a standalone `static constexpr auto` variable, Clang behaves as expected: No `VLA` is used and no warning for folding a constant is emitted.

Please let me know if you require more information. This is my first issue, I hope it is clear :)

Thanks for providing and maintaining this amazing tool!

**Comment by shafik** [comment]
CC @AaronBallman @Fznamznon

**Comment by Fznamznon** [comment]
Hmm, AFAIK 
```
static constexpr struct
{   int SHIFT;
    int ENTRIES;
}
PageTableFormat = {.SHIFT = 9, .ENTRIES = (1ULL << PageTableFormat.SHIFT)};
```
is an underspecified declaration, which clang doesn't officially support yet (https://clang.llvm.org/c_status.html )
I suppose with applying https://github.com/llvm/llvm-project/pull/79845/files , clang would reject. As well as latest gcc.

Though, the latest C working draft that I could find says that

> A declaration such that the declaration specifiers contain no type specifier or that is declared with
constexpr is said to be underspecified. If such a declaration is not a definition, if it declares no or more
than one ordinary identifier, if the declared identifier already has a declaration in the same scope,
if the declared entity is not an object, or if anywhere within the sequence of tokens making up the
declaration identifiers that are not ordinary are declared, the behavior is implementation-defined.

with a side note that

> It is recommended that implementations that accept such declarations follow the semantics of the corresponding feature
in ISO/IEC 14882.

**Comment by AaronBallman** [comment]
Yeah, that specific code should be rejected as an underspecified declaration in C23. A slight change to:
```
struct PTF
{   int SHIFT;
    int ENTRIES;
};

static constexpr struct PTF
PageTableFormat = {.SHIFT = 9, .ENTRIES = (1ULL << PageTableFormat.SHIFT)};
```
is also not valid because C23 6.7.11p24: The evaluations of the initialization list expressions are indeterminately sequenced with respect to
one another and thus the order in which any side effects occur is unspecified.

So there's no guarantee that `.SHIFT` has been initialized prior to using it in `PageTableFormat.SHIFT`. But if we fix that, then we get to the crux of the bug, which is: https://godbolt.org/z/oYdhjx849

All of the warnings Clang emits there seem incorrect because `PageTableFormat` is a named constant (C23 6.6p7) and those are allowed in an ICE (p8). It looks like we're missing that logic in Clang.

**Comment by florianmarkusse** [comment]
Pardon my ignorance for this but trying to comprehend the comments. 

## Case 1

```
struct PTF
{   int SHIFT;
    int ENTRIES;
};

static constexpr struct PTF
PageTableFormat = {.SHIFT = 9, .ENTRIES = (1ULL << 9)};
```

In above scenario, according to the C23 standard, using `PTF.ENTRIES` as a size specifier to declare an array should result in a constant array in the same way as declaring `static constexpr int TEST = 5;` and using that variable.

While,

## Case 2

```
struct PTF
{   int SHIFT;
    int ENTRIES;
};

static constexpr struct PTF
PageTableFormat = {.SHIFT = 9, .ENTRIES = (1ULL << PageTableFormat.SHIFT)};
```
and attempting to do the same is underspecified because the order in which the expressions are initialized in the struct is indeterminate. Hence, the standard says it is implementation-defined. 

---
Currently, Clang emits false-positives in Case 1. Are these emitted warnings false positives? Or is it actually using a `VLA` while it should be using a constant array according to the C23 specification?


Is that a correct understanding?

**Comment by AaronBallman** [comment]
> Pardon my ignorance for this but trying to comprehend the comments.

No need for apologies, this is complicated language lawyering stuff. :-)

> ## Case 1
> 
> ```
> struct PTF
> {   int SHIFT;
>     int ENTRIES;
> };
> 
> static constexpr struct PTF
> PageTableFormat = {.SHIFT = 9, .ENTRIES = (1ULL << 9)};
> ```
> 
> In above scenario, according to the C23 standard, using `PTF.ENTRIES` as a size specifier to declare an array should result in a constant array in the same way as declaring `static constexpr int TEST = 5;` and using that variable.

Correct.

> While,
> ## Case 2
> 
> ```
> struct PTF
> {   int SHIFT;
>     int ENTRIES;
> };
> 
> static constexpr struct PTF
> PageTableFormat = {.SHIFT = 9, .ENTRIES = (1ULL << PageTableFormat.SHIFT)};
> ```
> 
> and attempting to do the same is underspecified because the order in which the expressions are initialized in the struct is indeterminate. Hence, the standard says it is implementation-defined.

It's underspecified because it uses the `constexpr` storage class specifier. C23's wording is: A declaration such that the declaration specifiers contain no type specifier or that is declared with constexpr is said to be underspecified. (So that also covers use of `auto` as in `auto x = 12;` where the type is inferred.)

Separately, the initialization rules say that initialization is indeterminately sequenced, so they can be evaluated in any order. That's not even implementation-defined behavior, that's unspecified behavior where you are guaranteed to get *some* resulting value, but you could get a different value on each evaluation of the initializer.

> Currently, Clang emits false-positives in Case 1. Are these emitted warnings false positives? Or is it actually using a `VLA` while it should be using a constant array according to the C23 specification?
> 
> Is that a correct understanding?

In Case 1, the `-Wvla` diagnostics are false positives because the use should result in an integer constant expression. GCC gets this correct, Clang gets it wrong: https://godbolt.org/z/9z3ezET8h

**Comment by florianmarkusse** [comment]
Thanks for clarifying!

---

And the term 'underspecified' here infers that a compiler can decide itself on if and how to implement this behavior, correct?

> Separately, the initialization rules say that initialization is indeterminately sequenced, so they can be evaluated in any order. That's not even implementation-defined behavior, that's unspecified behavior where you are guaranteed to get some resulting value, but you could get a different value on each evaluation of the initializer.

I wonder how 
```
struct PTF
{   int SHIFT;
    int ENTRIES;
};

static constexpr struct PTF
PageTableFormat = {.SHIFT = 9, .ENTRIES = (1ULL << PageTableFormat.SHIFT)};
```

could be evaluated other than `SHIFT` first and then use that for `ENTRIES` but I can see scenarios where this may not be so cut-and-dried :P.

---

> In Case 1, the -Wvla diagnostics are false positives because the use should result in an integer constant expression. GCC gets this correct, Clang gets it wrong: https://godbolt.org/z/9z3ezET8h

I am asking specifically about the false positive because the identically-named label has the description that a warning fires when it should not. In the example you give, I agree.  

However, in: https://godbolt.org/z/naGr8PofE, I can see that the IR for `testFunction2` and `testFunction3` differs even though they should be the same, right? I also see that the IR for a function where I purposefully use a VLA, `vlaOnPurpose`, is again different.

I am not an expert or even knowledgeable in LLVM IR, I just found it odd that `testFunction2` and `testFunction3` end up with different IR based on my fledgling understanding of the standard. Can you shed some light on this?

**Comment by AaronBallman** [comment]
> Thanks for clarifying!
> 
> And the term 'underspecified' here infers that a compiler can decide itself on if and how to implement this behavior, correct?

Not quite; it's a term of art in the C standard. The paper proposing the idea is: https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3006.htm (there were some later changes based on national body comments, so the final C23 text is slightly different).

> > Separately, the initialization rules say that initialization is indeterminately sequenced, so they can be evaluated in any order. That's not even implementation-defined behavior, that's unspecified behavior where you are guaranteed to get some resulting value, but you could get a different value on each evaluation of the initializer.
> 
> I wonder how
> 
> ```
> struct PTF
> {   int SHIFT;
>     int ENTRIES;
> };
> 
> static constexpr struct PTF
> PageTableFormat = {.SHIFT = 9, .ENTRIES = (1ULL << PageTableFormat.SHIFT)};
> ```
> 
> could be evaluated other than `SHIFT` first and then use that for `ENTRIES` but I can see scenarios where this may not be so cut-and-dried :P.
> 
> > In Case 1, the -Wvla diagnostics are false positives because the use should result in an integer constant expression. GCC gets this correct, Clang gets it wrong: https://godbolt.org/z/9z3ezET8h
> 
> I am asking specifically about the false positive because the identically-named label has the description that a warning fires when it should not. In the example you give, I agree.
> 
> However, in: https://godbolt.org/z/naGr8PofE, I can see that the IR for `testFunction2` and `testFunction3` differs even though they should be the same, right? I also see that the IR for a function where I purposefully use a VLA, `vlaOnPurpose`, is again different.
> 
> I am not an expert or even knowledgeable in LLVM IR, I just found it odd that `testFunction2` and `testFunction3` end up with different IR based on my fledgling understanding of the standard. Can you shed some light on this?

I would have expected the IR for `testFunction3` to be identical to that of `testFunction2`, where the size is statically known and we didn't need to generate any loads or stores. We didn't *really* make a VLA because we did `%4 = alloca i32, i64 10, align 16, !dbg !38` because we already knew that we needed 10 elements.

**Comment by florianmarkusse** [comment]
> I would have expected the IR for testFunction3 to be identical to that of testFunction2, where the size is statically known and we didn't need to generate any loads or stores. We didn't really make a VLA because we did %4 = alloca i32, i64 10, align 16, !dbg !38 because we already knew that we needed 10 elements.

That was my thinking too, good to see it validated ^^.

Thanks for your patience in explaining all this! Is anything else necessary from my side to potentially have this issue fixed?
